torch/autograd/graph.py

```diff
@@ -112,8 +117,12 @@ class save_on_cpu():
         >>> # all intermediary tensors are released (deleted) after the call to backward
 
     """
-    def __init__(self, pin_memory=False):
+    def __init__(self, pin_memory=False, non_blocking_copy_to_cpu=False):
```

Comment:
This is redundant with the `pin_memory` argument no? We should have only one of them.

Reply:
> Also @ngimel is there any reason not to do the copies async here by default? The engine does restore the ambient stream so I can't see any way for this to go wrong?

Async cpu copies require synchronizations somewhere. Otherwise, you'll call `pack_to_cpu`, `unpack_from_cpu` (all asynchronous), return from those calls, cpu tensor goes out of scope, that memory is reused, by the time actual copy queued by `unpack_from_cpu` happens it contains garbage.

Reply:
@albanD I'm guessing we need to keep the `pin_memory` arg for BC. 

Although if we only have `pin_memory`, there will not be a flag to control whether we move the tensor to cpu with `non_blocking=True` in `pack_to_cpu`. Although, it was mentioned that non-blocking copy to CPU creates an automatically pinned memory tensor, so maybe the code can simply be something like:

```
def pack_to_cpu(tensor):
    oacked = tensor.to("cpu", non_blocking=pin_memory)
    return (tensor.device, packed)
```

and we can keep only the pin_memory argument? 

Regarding the synchronization issues, my understanding is that the CPU tensor will not go out of scope as it is saved by autograd so it won't be deleted before `unpack_from_cpu` has executed.

Reply:
@rohan-varma it is not, but it will be cleared after that backward function is executed.
But at that time, the async copy to the GPU might not have happened yet so it is not valid to release this CPU Tensor.

@ngimel isn't this issue what https://github.com/pytorch/pytorch/blob/3c2ae2b47cc859847e7915a406a7484401cba35e/aten/src/ATen/native/cuda/Copy.cu#L227 is trying to avoid?

Reply:
It would prevent memory deallocation, but it wouldn't prevent tensor being filled with something else:
```
In [16]: a=torch.empty(10, pin_memory=True); a.fill_(1); b=torch.empty(10, device="cuda"); torch.cuda._sleep(5000000000); b.copy_(a, non_blocking=True); a.fil
    ...: l_(5); print(b);
tensor([5., 5., 5., 5., 5., 5., 5., 5., 5., 5.], device='cuda:0')
```
Here, we would expect `b` to be full of ones, because that was `a` contents at the time the copy was submitted, but instead it's filled with `5`. 
My point is, everywhere there's async operation, pytorch doesn't (and cannot) know how to do the right thing, and its authors job to either do necessary synchronizations, or leave comments why unfavorable sequence of events will never happen. Just changing blocking copies into non-blocking is not the right thing to do (even though it may still have the correct behavior).